### PR TITLE
Upgrade cypress and cucumber libraries

### LINKED ISSUE #6975: Upgrade cypress and cucumber test libraries
### Which dependency needs to be upgraded, and why?

Cypress and cucumber. To be consistent with OSSMC where new integration tests are going to use latest versions of cypress and cucumber.

### What is the current version of the dependency?
Cypress: 10.7.0
Cucumber: 12.2.0

### Which version to upgrade to?
Cypress: 13.6.1
Cucumber: 19.2.0

### Changelog link
Cypress: https://docs.cypress.io/guides/references/changelog
Cucumber: https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/CHANGELOG.md

### PR DESCRIPTION
**Pull request by ferhoyos**
**Describe the change**

Upgrade cypress and cucumber libraries to latest versions:
Cypress: 13.6.1
Cucumber: 19.2.0

`And` and `But` elements are deprecated in latest Cypress version, so I have replaced them with `When` or `Then`. The rest of the changes are lint fixes.

**Issue reference**

Closes #6975

### CODE REVIEW

**Comment by ScriptingShrimp** [comment]
@ferhoyos this is great! Can we also backport this into 1.73, so our "LTS" branch does not diverge too much in terms of support? WDYT?

**Comment by ferhoyos** [comment]
Sure, when the PR is merged I can do the backport.

File: frontend/cypress/integration/common/sidecar_injection.ts [hunk 1]
```diff
@@ -237,25 +237,28 @@ Given('a workload with override configuration for automatic sidecar injection',
...
-When('I visit the overview page', function () {
+When('I visit the overview page', () => {
   cy.visit('/console/overview?refresh=0');
   cy.contains('Inbound traffic', { matchCase: false }); // Make sure data finished loading, so avoid broken tests because of a re-render
 });
 
 // Only works for single cluster.
 When('I override the default automatic sidecar injection policy in the namespace to enabled', function () {
-  cy.request('GET', '/api/status').should(response => {
+  cy.request('GET', '/api/status').then(response => {
```
Comment (nrfox) [inline frontend/cypress/integration/common/sidecar_injection.ts:247]: Same question about `should` vs. `then` for the assertions on this page.

File: frontend/cypress/integration/common/workload_logs.ts [hunk 2]
```diff
@@ -1,88 +1,96 @@
...
-    cy.get('[type=checkbox]:checked').uncheck();
+    cy.get('[type=checkbox]').uncheck();
+
     cy.get(`input#container-${containerName}`).check();
   });
 });
 
 When('I enable visualization of spans', () => {
   cy.get('[data-test=workload-logs-pod-containers]').within(() => {
-    cy.get('[type=checkbox]:checked').uncheck();
```
Comment (nrfox) [inline frontend/cypress/integration/common/workload_logs.ts:37]: This is a different assertion now?

File: frontend/cypress/support/commands.ts [hunk 3]
```diff
@@ -70,6 +70,23 @@ declare namespace Cypress {
...
+      kialiToken = cookie.value;
+      haveCookie = true;
+    });
+};
+
+const getCsrfToken = (response: Cypress.Response<any>): string => {
+  const $html = Cypress.$(response.body);
+  const csrf = $html.find('input[name=csrf]').val();
+  return csrf.toString();
+};
```
Comment (nrfox) [inline frontend/cypress/support/commands.ts:89]: Just curious why the change in how the cookie is gotten? i.e. going from `cy.getCookie('csrf')` to getting it from the html?

File: frontend/cypress/support/commands.ts
Comment (nrfox) [inline frontend/cypress/support/commands.ts:1]: What was the need for changing the login flow? The Kiali token is needed for something?

**Review by nrfox — COMMENTED** [review]
Couple questions otherwise LGTM. Thanks for doing this. Updates like this can be large and tedious.

File: frontend/cypress/support/commands.ts [hunk 4]
```diff
@@ -174,39 +191,39 @@ Cypress.Commands.add('login', (username: string, password: string) => {
...
 
-        cy.getCookie('kiali-token-aes', { timeout: 15000 })
-          .should('exist')
-          .then(() => {
-            haveCookie = true;
-          });
+          preserveKialiToken();
+        }
       } else if (auth_strategy === 'token') {
         cy.exec('kubectl -n istio-system create token citest').then(result => {
```
[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply frontend/cypress/support/commands.ts:228]: I was talking to @ferhoyos and this is somehow broken. I was not able find easy solution and local execution does not work against OCP only. Therefore I don't want to block this PR by this issue, we can fix or refactor auth mechanism later (cc @nrfox )

File: frontend/cypress/plugins/setup.ts [hunk 5]
```diff
@@ -4,12 +4,14 @@
 */
 import get from 'axios';
 
-export async function getAuthStrategy(url: string) {
+export const getAuthStrategy = async (url: string): Promise<any> => {
```
[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply frontend/cypress/plugins/setup.ts:7]: will be resolved in favor of https://github.com/kiali/kiali/pull/6980#discussion_r1447639208

File: frontend/cypress/support/commands.ts (hunk 3 above)
Comment (ferhoyos) [reply frontend/cypress/support/commands.ts:89]: I don't know why but `cy.getCookie('csrf')` does not work with latest version of Cypress (the method just hangs up throwing a timeout). I found and alternative way to get csrf token from official cypress examples (https://www.github.com/cypress-io/cypress-example-recipes/blob/decceecf8ff3a60e2b2f2dfa2461b4c3672949ff/examples/logging-in__csrf-tokens/cypress/e2e/logging-in-csrf-tokens-spec.cy.js#L69)

File: frontend/cypress/integration/common/hooks.ts [hunk 6]
```diff
@@ -64,40 +70,49 @@ function install_demoapp(demoapp: string): void {
       });
     }
   });
-}
+};
 
-Before({ tags: '@gateway-api' }, async () => {
```
[the remark this answers is not in the document]
Comment (ferhoyos) [reply frontend/cypress/integration/common/hooks.ts:69]: This is the error Cypress provides when you have async:
```
Cucumber preprocessor detected that you returned a native promise from a function handler, this is not supported. Using async / await is generally speaking not supported when using Cypress, period, preprocessor or not.
```
It seems that current cucumber preprocess does not allow async functions in Before tags...

File: frontend/cypress/integration/common/istio_config.ts [hunk 7]
```diff
@@ -438,27 +457,32 @@ Then('user sees {string}', (sees: string) => {
...
     cy.contains(object).click();
   });
+
   const page = `/istio/new/${object}`;
   cy.url().should('include', page);
 });
 
 Then('the user can create a {string} K8s Istio object', (object: string) => {
-  cy.request('GET', '/api/config').should(response => {
+  cy.request('GET', '/api/config').then(response => {
```
[the remark this answers is not in the document]
Comment (ferhoyos) [reply frontend/cypress/integration/common/istio_config.ts:470]: Exactly. A retry should be performed inside a callback. This was the Cypress error I found on upgrade:
```
CypressError: `cy.should()` failed because you invoked a command inside the callback. `cy.should()` retries the inner function, which would result in commands being added to the queue multiple times. Use `cy.then()` instead of `cy.should()`, or move any commands outside the callback function.
```

File: frontend/cypress/integration/common/workload_logs.ts (hunk 2 above)
Comment (ferhoyos) [reply frontend/cypress/integration/common/workload_logs.ts:37]: Previously only checked checkboxes were unchecked. Now all checkboxes are unchecked, for the cypress test purpose is the same...

This was the Cypress error:
```
CypressError: Timed out retrying after 40100ms: `cy.uncheck()` failed because the page updated while this command was executing. Cypress tried to locate elements based on this query:
```
I think the system gets unstable because unchecks are run in parallel, so it is hard for cypress to determine which checkboxes are checked if at the same time some of them are being unchecked.

File: frontend/cypress/support/commands.ts
Comment (ferhoyos) [reply frontend/cypress/support/commands.ts:1]: `Cypress.Cookies.preserveOnce` is deprecated in latest version of cypress, so I had to find an alternative way to store Kiali auth token in order not to login for every test case.

https://github.com/kiali/kiali/pull/6980/files/0899461e9a76dff38b88bc6c883a205f72f48652#diff-9596564ea62b2b4aa7b6cf65a4457cda4bca7f0cb390251544b9a4e9b271003aL38

File: frontend/cypress/support/commands.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrfox) [reply frontend/cypress/support/commands.ts:228]: Hmmm I think this still works in CI?

File: frontend/cypress/support/commands.ts
Comment (nrfox) [reply frontend/cypress/support/commands.ts:1]: Created: https://github.com/kiali/kiali/issues/7018 as a followup

**Review by nrfox — APPROVED** [review]

File: frontend/cypress/support/commands.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply frontend/cypress/support/commands.ts:228]: It works on GH actions as well as in Jenkins

**Review by ScriptingShrimp — APPROVED** [review]
LGTM

File: frontend/cypress/perf/Graphload.spec.ts [hunk 8]
```diff
@@ -95,26 +102,13 @@ describe('Performance tests', () => {
...
-              data.namespaces +
-              '&duration=' +
-              data.duration +
-              '&refresh=' +
-              data.refresh +
-              '&layout=' +
-              data.layout +
-              '&namespaceLayout=' +
-              data.namespaceLayout
+            `/console/graph/namespaces?traffic=${data.traffic}&graphType=${data.graphType}&namespaces=${data.namespaces}&duration=${data.duration}&refresh=${data.refresh}&layout=${data.layout}&namespaceLayout=${data.namespaceLayout}`
```
[the remark this answers is not in the document]
Comment (ferhoyos) [reply frontend/cypress/perf/Graphload.spec.ts:111]: String template eslint rule applied (https://www.github.com/kiali/kiali/issues/6736)

[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply frontend/cypress/perf/Graphload.spec.ts:111]: @matejnesuta FYI
